File: rest/src/main/java/io/disquark/rest/resources/webhook/ModifyWebhookWithToken.java
package io.disquark.rest.resources.webhook;

import static io.disquark.rest.util.Variables.variables;

import java.util.Optional;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

import io.disquark.immutables.ImmutableUni;
import io.disquark.nullableoptional.NullableOptional;
import io.disquark.nullableoptional.jackson.NullableOptionalFilter;
import io.disquark.rest.jackson.ImageDataSerializer;
import io.disquark.rest.request.AbstractRequestUni;
import io.disquark.rest.request.Endpoint;
import io.disquark.rest.request.Request;
import io.disquark.rest.resources.Snowflake;
import io.smallrye.mutiny.subscription.UniSubscriber;
import io.vertx.core.http.HttpMethod;
import io.vertx.mutiny.core.buffer.Buffer;

import org.immutables.value.Value.Redacted;

@ImmutableUni
abstract class ModifyWebhookWithToken extends AbstractRequestUni<Webhook> {

    @JsonIgnore
    public abstract Snowflake webhookId();

    @JsonIgnore
    public abstract String webhookToken();

    public abstract Optional<String> name();

    @Redacted
    @JsonSerialize(contentUsing = ImageDataSerializer.class)
    @JsonInclude(value = Include.CUSTOM, valueFilter = NullableOptionalFilter.class)
    public abstract NullableOptional<Buffer> avatar();

    @Override
    public void subscribe(UniSubscriber<? super Webhook> downstream) {
        requester().request(asRequest()).flatMap(res -> res.as(Webhook.class)).subscribe().withSubscriber(downstream);
    }

    @Override
    public Request asRequest() {
        return Request.builder()
                .endpoint(Endpoint.create(HttpMethod.PATCH, "/webhooks/{webhook.id}/{webhook.token}", false))
                .variables(variables("webhook.id", webhookId().getValue(), "webhook.token", webhookToken()))
                .body(this)
                .build();
    }
}

File: rest/kotlin/src/main/kotlin/json/webhook/ModifyWebhookWithToken.kt
package io.disquark.rest.kotlin.json.webhook

import io.disquark.rest.json.Snowflake
import io.disquark.rest.json.webhook.ModifyWebhookWithTokenUni
import io.disquark.rest.kotlin.nullableoptional.toNullableOptional
import io.disquark.rest.request.Requester
import io.vertx.mutiny.core.buffer.Buffer
import java.util.*

class ModifyWebhookWithToken(
    private val requester: Requester<*>,
    private val webhookId: Snowflake,
    private val webhookToken: String,
    var name: String? = null,
    var avatar: Optional<Buffer>? = Optional.empty(),
) {
    internal fun toUni(): ModifyWebhookWithTokenUni {
        return ModifyWebhookWithTokenUni.builder()
            .requester(requester)
            .webhookId(webhookId)
            .webhookToken(webhookToken)
            .name(name)
            .avatar(avatar.toNullableOptional())
            .build()
    }
}
